### ISSUE TITLE
Ability to see how long an action's been pressed

### ISSUE THREAD

**Issue by romlok** [godotengine/godot#21005]
**Godot version:**
Godot 3.0.6 (Steam)

**Feature request:**

```
float Input.get_action_duration(String action)
    Returns the amount of time elapsed since the given action last became pressed.
```

**Background:**

Sometimes it is desirable to have the same input perform different actions depending on how long it's been held down.
It my most recent case, I have left mouse button select a tile if clicked, but scroll the map when dragged. Currently, my code has something like this:

```
func input(event):
    if event.is_action_pressed("select"):
        click_timer = get_tree().create_timer(0.3)
        
    elif event.is_action_released("select"):
        if click_timer != null and click_timer.time_left <= 0:
            # Drag events are dealt with in the camera's _unhandled_input
            return
        # Select the clicked tile
        select_tile()
```
But I would very much like to be able to do something more like:
```
func input(event):
    if event.is_action_released("select") and Input.get_action_duration("select") < 0.3:
        select_tile()
```

**Comment by Xrayez** [comment]
Currently the engine stores internal physics and idle frames counts within each actions' states with `Engine.get_physics/idle_frames()`.

So the only thing which is needed is computing the difference between current and cached frames, and converting the number of frames to seconds, and of course exposing it as `get_action_duration()`.

The alternative is just letting the user to fetch the physics/idle frame timestamp with `Input.get_action_physics/idle_frame` methods with combination of `Engine.get_physics/idle_frames()` which are also available in GDScript in since 3.2: #35092, but it might be too low-level.

See also godotengine/godot-proposals#104 / #35240 which aims to cover many inaccessible input states like timestamps.

The feature would be likely helpful for fighting games and similar games where a player can make a combo move over a period of time, see godotengine/godot-proposals#100.

**Comment by girng** [comment]
You might be able to do this by simply setting a  `timestamp` to a variable, and tracking it when the action is pressed

**Comment by Xrayez** [comment]
@ca3games you've made a proposal regarding input buffers some time ago which was basically rejected: godotengine/godot-proposals#100. I wonder whether you'd find #36460 useful for your use cases.

@LikeLakers2 [presented](https://github.com/godotengine/godot-proposals/issues/100#issuecomment-534818031) what seems to be an acceptable compromise between too-specific feature towards something which an engine can lend a hand, and #36460 seems to implement very similar feature:

> Then you could call something like Input.is_action_recently_pressed("some_action_name", time_in_seconds) (returns true if the action has been pressed within that time) or Input.get_action_last_press("some_action_name") (returns a float representing the amount of time since the action was last pressed). I wonder if something like this could be an acceptable solution?

By the way, `Input.is_action_just_pressed/released` can also be implemented via script currently since 3.2 as you can fetch timestamps: #35092, ~~so those can be removed.~~ 😛

**Comment by clayjohn** [comment]
Feature and improvement proposals for the Godot Engine are now being discussed and reviewed in a dedicated [**Godot Improvement Proposals (GIP)**](https://github.com/godotengine/godot-proposals) ([godotengine/godot-proposals](https://github.com/godotengine/godot-proposals)) issue tracker. The GIP tracker has a detailed issue template designed so that proposals include all the relevant information to start a productive discussion and help the community assess the validity of the proposal for the engine.

The main ([godotengine/godot](https://github.com/godotengine/godot)) tracker is now solely dedicated to bug reports and Pull Requests, enabling contributors to have a better focus on bug fixing work. Therefore, we are now closing all older feature proposals on the main issue tracker.

If you are interested in this feature proposal, please [**open a new proposal on the GIP tracker**](https://github.com/godotengine/godot-proposals/issues/new/choose) following the given issue template (after checking that it doesn't exist already). Be sure to reference this closed issue if it includes any relevant discussion (which you are also encouraged to summarize in the new proposal). Thanks in advance!
